README
HTML::Template::Ex
=============================

	The Perl code is executed in HTML::Template.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


COPYRIGHT

	Copyright 2004 Bee Flag .lp <http://beeflag.com/>, All Rights Reserved.

	This program is free software; you can redistribute it and/or modify it
	under the same terms as Perl itself.


AUTHOR

	Masatoshi Mizuno, <mizuno@beeflag.com>



Changes
Revision history for Perl extension .

0.06
	- lost property.

0.05
	- bug fix.
	- HTML::Template::Ex::Filter was enclosed.

0.04
	- Addition of '_get_ident_id' internal function.

	- It was made to encode only when the character set was able to be picked up from the template.
	  It is related to constructor's option 'auto_encoder'.

	- It came to be able to sweeten the check on the syntax of EX-Code.
	  It is related to constructor's option 'no_strict_exec'.

	- Trifling bug fix.

0.03
	- When the template was read, the character-code can have been changed.
	  It is related to constructor's option 'encoder'.

	- Charset of the read template was obtained.
	  The 'charset' method of the relation is added.

	- EX-Code was evaluated at output.

	- The problem concerning dispatching priority was canceled.
	  It is processed in order of <TMPL_EX> appearing from the template regardless
	  of the presence of <TMPL_INCLUDE *NAME>. 

0.02
	- The module is not updated.
	  Because an attached file was too rude, it corrected it.

0.01  Wed Jul 20 16:33:07 2005
	- original version; created by h2xs 1.21 with options
		-A HTML::Template::Ex


test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

#########################

# change 'tests => 1' to 'tests => last_test_to_print';

use Test;
BEGIN { plan tests => 1 };
use HTML::Template::Ex;
ok(1); # If we made it this far, we're ok.

#########################

# Insert your test code below, the Test module is use()ed here so read
# its man page ( perldoc Test ) for help writing this test script.



example/lib/myconfig.pm
package myconfig;
#
# Copyright 2004 Bee Flag, Corp. All Rights Reserved.
# Masatoshi Mizuno <mizuno@beeflag.com>
#
# $Id: myconfig.pm,v 1.2 2005/07/28 09:55:12 Lushe Exp $
#
use strict;

my $Prefix= '/home/exbase';
my $VRsite= "$Prefix/vrsite";

my %ConvertCodes= (
 'utf-8'      => 'utf8',
 'shift_jis'  => 'sjis',
 'x-sjis'     => 'sjis',
 'jis'        => 'jis',
 'iso-2022-jp'=> 'jis',
 );

my %C= (
 language => 'ja',
 charset  => 'euc-jp',
 title    => 'ExBase',
 siteRoot => '/',
 indexName=> 'index.html',
 options  => {
#  encoder=> 1,
  cache=> 1,
  path => ["$Prefix/plugins", "$Prefix/template"],
  },
 deflateOk  => 0,
 deflateSize=> 2048,
 #
 prefix   => $Prefix,
 VRsite   => $VRsite,
 convCodes=> \%ConvertCodes,
 );

sub config { \%C }

1;




example/lib/ExBase/Core.pm
package ExBase::Core;
#
# Copyright 2004 Bee Flag, Corp. All Rights Reserved.
# Masatoshi Mizuno <mizuno@beeflag.com>
#
# $Id: Ex.pm,v 1.1 2005/07/21 15:13:10 Lushe Exp $
#
use strict;
use CGI;
use Jcode;
use FileHandle;
use HTML::Template::Ex;

our $VERSION = q$0.04$;
our $REVISION= q$Id: base.cgi,v 1.2 2005/07/21 16:12:28 Lushe Exp $;

sub REVISION { $REVISION }
sub eucConv  { defined($_[1]) ? Jcode->new($_[1])->euc: "" }
sub cgi { $_[0]->{cgi} }

sub run {
	my($class, $conf, $apr)= @_;
	my $base= bless $conf, $class;
	my $cgi = $base->{cgi}= CGI->new($apr);
	$base->{options}{encoder}
	  and $base->{options}{encoder}= sub { $base->eucConv(@_) };
	my($path, $file);
	$conf->{___ApacheHandler} ? do {
		($path, $file)= $base->pathCuter($ENV{REQUEST_URI});
	 }: do {
		($path, $file)=
		 $cgi->param('path') ? $base->pathCuter($cgi->param('path')):
		 ($ENV{QUERY_STRING} && $ENV{QUERY_STRING}=~/^([^\&\;]+)/)
		 ? $base->pathCuter($1): ('/', $base->{indexName});
	 };
	$file=~s{\.htm$} [.html];
	(-e "$base->{VRsite}$path$file" && -f _)
	 ? do { $base->execute($cgi, $path, $file) }
	 : do { print $cgi->header. q{<h1>404 File Not Found.</h1>} };
}
sub execute {
	my($base, $cgi, $path, $file)= @_;
	my $send= $base->{deflateOk} ? \&sendContent: \&sendContent;
	$base->{currentPage}= "$path$file";
	$base->{rewrite} ? do {
		$base->{baseRoot}  = $base->{siteRoot};
		$base->{currentUri}= $base->{currentPage};
	 }: do {
		$base->{baseRoot}  = "$base->{cgiPath}?$base->{siteRoot}";
		$base->{currentUri}= "$base->{cgiPath}?$base->{currentPage}";
	 };
	my $Vars= $cgi->Vars;
	while (my($name, $value)= each %$Vars) {
		$value=~tr/\t/ /;
		$value=~s{\r\n?} [\n]sg;
		$Vars->{$name}= $base->eucConv($value);
	}
	my @path= ("$base->{VRsite}$path", @{$base->{options}{path}});
	my %options= %{$base->{options}};
	$options{path}     = \@path;
	$options{filename} = $file;
	$options{associate}= [$cgi];
	my $tmpl;
	eval{ $tmpl= HTML::Template::Ex->new($base, \%options) };
	$@ ? do {
		$cgi->charset($base->{charset});
		print $cgi->header, "Internal Server Error: $@";
	 }: do {
		$base->{sendCharset}= $tmpl->charset || $base->{charset};
		$cgi->charset($base->{sendCharset});
		$tmpl->param({
		 title=> $base->{title},
		 cgiPath => $base->{cgiPath},
		 baseRoot=> $base->{baseRoot},
		 siteRoot=> $base->{siteRoot},
		 currentUri    => $base->{currentUri},
		 currentPage   => $base->{currentPage},
		 scriptVersion => sub { $base->scriptVersion },
		 supportURL    => 'http://exbase.luuu.net/',
		 publishCharset => $base->{sendCharset},
		 publishLanguage=> $base->{language},
		 });
		$send->($base, $cgi, $tmpl);
	 };
}
sub sendContent {
	my($base, $cgi, $tmpl)= @_;
	my $code= $base->{convCodes}{lc($base->{sendCharset})}
	|| do { return print $cgi->header, $tmpl->output };
	print $cgi->header, Jcode->new($tmpl->output, 'euc')->$code;
}
sub sendDeflateContent {
}
sub fileOpen {
	my $base= shift;
	my $file= shift || return 0;
	FileHandle->new($file) || 0;
}
sub pathCuter {
	my $base= shift;
	my $path= shift || return ("/", $base->{indexName});
	$path=~s/\?.+//;
	my($dir, $file)=
	  $path=~m{^(/.*?)([^/]+)$} ? ($1, $2): ($path, $base->{indexName});
	(($dir || '/'), $file);
}
sub scriptVersion { $REVISION=~/v\s+([\.\d]+)/; __PACKAGE__." V$1" }

1;



lib/HTML/Template/Ex/Filter.pm
package HTML::Template::Ex::Filter;
#
# Copyright (C) 2007 Bee Flag, Corp, All Rights Reserved.
# Masatoshi Mizuno E<lt>mizunoE<64>bomcity.comE<gt>
#
# $Id: Filter.pm 221 2007-02-22 13:54:45Z lushe $
#
use strict;
use warnings;

our $VERSION = '0.01';

sub set {
	{
	 format=> 'scalar',
	 sub=> sub {
		my $text= shift;
		$$text=~s{\[\%\s*([^\%]+)\s*\%\]} [ &filter($1) ]sge;
	  },
	 };
}
sub filter {
	local($_)= @_;
	return
	/^(?:var|\=)\s*(\S+)/i ? do {
		my $v= $1;
		$v=~/^(url|html)\s*\:(.+)/i
		  ? qq{<tmpl_var escape="$1" name="$2">}
		  : qq{<tmpl_var name=$v>};
	 }:
	/^ex(?:\s+(.+))?/i ? do {
		my($str, $option)= ($1, "");
		if ($str) {
			my($name, $escape)=
			  $str=~/^([^\:\"\']+)\s*\:\s*([^\s\"\']+)/ ? ($2, $1):
			  $str=~/^([^\:\"\']+)\:\s*$/ ? ("", $1):
			  $str=~/([^\s\"\']+)/ ? ($1, ""): ("", "");
			$option = qq{ escape="$escape} if $escape;
			$option.= qq{ name="$name"}    if $name;
		}
		qq{<tmpl_ex$option>};
	 }:
	/^set\s+([^\s\=]+)\s*\=\s*(.+)/i ? do {
		my($name, $value)= ($1, $2);
		$value=~s/(?:^[\"\']|[\"\']\s*$)//g;
		qq{<tmpl_set name="$name" value="$value">};
	 }:
	/^(if|unless)\s+(\S+)/i          ? do { qq{<tmpl_$1 name=$2>} }:
	/^else/i                         ? do { qq{<tmpl_else>} }:
	/^loop\s+(\S+)/i                 ? do { qq{<tmpl_loop name=$1>} }:
	/^(?:end_loop|\/\s*loop)/        ? do { qq{</tmpl_loop>} }:
	/^(?:include|\.\.\.?)\s*(\S+)/i  ? do { qq{<tmpl_include name=$1>} }:
	/^(?:end_|\/\s*)(if|unless|ex)/i ? do { qq{</tmpl_$1>} }:
	/^(?:\!|comment)/                ? do { "" }:
	do { qq{[% $_ %]} };
}

1;

__END__

=head1 NAME

HTML::Template::Ex::Filter - tmpl_tag filter for HTML::Template::Ex.

=head1 SYNOPSIS

  use HTML::Template::Ex;
  use HTML::Template::Ex::Filter;
  
  my $tmpl= HTML::Template::Ex->new(
    ...
    filter=> HTML::Template::Ex::Filter->set,
    );

  [%... include_template.tmpl %]
  
  [% ex %]
    my($self, $param)= @_;
    ...
    ..... ban, bo, bon.
    "";
  [% end_ex %]
  
  [% =param_name %]
  
  [% loop loop_param_name %]
    ...
  [% end_loop %]

=head1 DESCRIPTION

This module offers the filter to make the format of HTML::Template easy a little.

=head1 TAGS

=head2 [% ex %] ... [% end_ex %]

It corresponds to '<TMPL_EX> ... </TMPL_EX>'.

To specify the NAME attribute, as follows is done.

  [% ex param_name %] ...

In addition, delimit by the ESCAPE attribute and to the following

  [% ex html:param_name %]

NAME attribute a unspecified ESCAPE attribute must make the head and to the 
following

  [% ex html: %]

=head2 [% set PARAM_NAME="PARAM_VALUE" %]

It corresponds to '<TMPL_SET NAME="PARAM_NAME" VALUE="PARAM_VALUE">'.

=head2 [% =VAR_NAME %]

It corresponds to '<TMPL_VAR NAME="VAR_NAME">'.

The ESCAPE attribute does as follows.

  [% =html:VAR_NAME %]

=head2 [% if VAR_NAME %] ... [% else %] ... [% end_if %]

It corresponds to '<TMPL_IF NAME="VAR_NAME"> ... <TMPL_ELSE> ... </TMPL_IF>'.

=head2 [% unless VAR_NAME %] ... [% else %] ... [% end_unless %]

It corresponds to '<TMPL_UNLESS NAME="VAR_NAME"> ... <TMPL_ELSE> ... </TMPL_UNLESS>'.

=head2 [% loop ARRAY_NAME %] ... [% end_loop %]

It corresponds to '<TMPL_LOOP NAME="ARRAY_NAME"> ... </TMPL_LOOP>'.

=head2 [%... INCLUDE_TEMPLATE %]  or  [% include INCLUDE_TEMPLATE %]

It corresponds to '<TMPL_INCLUDE NAME="...">'.

=head2 [% ! COMMENT_STRING %]  or [% comment COMMENT_STRING %]

It is a comment. It is not reflected in the template.

=head1 SEE ALSO

L<Egg::Release>,

=head1 AUTHOR

Masatoshi Mizuno E<lt>mizunoE<64>bomcity.comE<gt>

=head1 COPYRIGHT

Copyright (C) 2007 by Bee Flag, Corp. E<lt>L<http://egg.bomcity.com/>E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/HTML/Template/Ex.pm
package HTML::Template::Ex;
#
# Copyright (C) 2007 Bee Flag, Corp, All Rights Reserved.
# Masatoshi Mizuno E<lt>mizunoE<64>bomcity.comE<gt>
#
# $Id: Ex.pm 225 2007-02-22 15:11:05Z lushe $
#
use strict;
use warnings;
use base qw/HTML::Template/;
use Carp qw/croak/;
use Digest::MD5 qw/md5_hex/;

our $VERSION = '0.06';

my $ErrstrStyle= <<END_OF_STYLE;
padding    : 5px;
background : #004858;
color      : #FFF;
font-size  : 13px;
END_OF_STYLE

sub initStyle { $ErrstrStyle= $_[1] }

my $GetCharSetRegix=
 qr{<meta.+?content=[\'\"]text/html\s*\;\s*charset=([A-Za-z0-9\-_]+)[\'\"].*?/?\s*>};

sub new {
	my $class= shift;
	my $base = shift || HTML::Template::Ex::DummyObject->new;
	my %opt  = $_[0] ? ($_[1] ? @_: %{$_[0]})
	  : croak __PACKAGE__.'::new: I want argument.';

	for (
	  [qw{ strict 0 }],
	  [qw{ file_cache 0 }],
	  [qw{ global_vars 1 }],
	  [qw{ shared_cache 0 }],
	  [qw{ die_on_bad_params 0 }],
	  ) {
		$opt{$_->[0]}= $_->[1] if exists($opt{$_->[0]});
	}

	my(%param, %mark, %order, %temp);
	$opt{_ex_base_object}= $base;
	$opt{_ex_params}     = \%param;
	$opt{_ex_orders}     = \%order;
	$opt{_ex_mark}       = \%mark;
	if (exists($opt{filter})) {
		if (ref($opt{filter}) eq 'CODE') {
			$opt{filter}= [{ format=> 'scalar', sub=> $opt{filter} }];
		} elsif (ref($opt{filter}) eq 'HASH') {
			$opt{filter}= [$opt{filter}];
		} elsif (ref($opt{filter}) ne 'ARRAY') {
			croak __PACKAGE__.q{::new: Bad format for 'filter'};
		}
	}
	if ($opt{setup_env}) {
		$param{"env_$_"}= sub { $ENV{$_} || "" } for keys %ENV;
	}
	my $filter= $opt{exec_off}
	 ? sub { &_offFilter(\%param, @_) }
	 : sub { &_exFilter($base, \%opt, \%temp, @_) };
	push @{$opt{filter}}, { format=> 'scalar', sub=> $filter };
	my $self= HTML::Template::new($class, %opt);
	$opt{cache} and $self->{_ex_charset}= pop @{$self->{parse_stack}} || "";
	$self;
}
sub output {
	my($self)= @_;
	my $parse_stack= $self->{parse_stack};
	my $options    = $self->{options};
	my($ex_mark, $ex_param, $ex_order);
	if ($options->{cache}) {
		$ex_mark = pop @$parse_stack;
		$ex_param= pop @$parse_stack;
		$ex_order= pop @$parse_stack;
	} else {
		$ex_mark = $options->{_ex_mark}   || {};
		$ex_param= $options->{_ex_params} || {};
		$ex_order= $options->{_ex_orders} || {};
	}
	HTML::Template::param($self, $ex_mark);
	my $base = $options->{_ex_base_object};
	my %param= %$ex_param;

	$options->{no_strict_exec} and do { no strict };  ## no critic
	my $cnt;
	for my $v (@$parse_stack) {
		next if (ref($v) ne 'HTML::Template::VAR' || ! $$v);
		my $hash= $ex_order->{$$v} || next;
		++$cnt;
		my $result;
		eval{ $result= $hash->{function}->($base, \%param) };
		if (my $err= $@) {
			require HTML::Entities;
			require Devel::StackTrace;
			my $st= Devel::StackTrace->new( no_refs=> 1 );
			my $error;
			$error.= $_->filename. ': '
			      .  $_->line. "\n" for $st->frames;
			$error = HTML::Entities::encode_entities(
			  "<TMPL_EX( $cnt )> - error:\n $err \n stack trace:\n $error",
			  q{\\\<>&\"\'},
			  );
			$error=~s{\r?\n} [<br />\n]sg;
			$param{$$v}= qq{<div style="$ErrstrStyle">$error</div>};
		} else {
			$param{$$v}= ref($result) eq 'ARRAY' ? ""
			           : $hash->{hidden} ? "": ($result || "");
			$param{$hash->{key_name}}= $result if $hash->{key_name};
		}
	}

	HTML::Template::param($self, \%param);
	my $result= HTML::Template::output(@_);
	if ($options->{cache}) {
		push @$parse_stack, $ex_order;
		push @$parse_stack, $ex_param;
		push @$parse_stack, $ex_mark;
		push @$parse_stack, ($self->{_ex_charset} || "");
	}
	$result;
}
sub charset { $_[0]->{_ex_charset} || "" }

sub _call_filters {
	my($self, $html)= @_;
	$self->{_ex_charset}= $1 if $$html=~m{$GetCharSetRegix}i;
	if ($self->{options}{auto_encoder}) {
		$self->{options}{auto_encoder}->($html) if $self->{_ex_charset};
	} elsif ($self->{options}{encoder}) {
		$self->{options}{encoder}->($html);
	}
	HTML::Template::_call_filters(@_);
}
sub _exFilter {
	my($base, $opt, $temp, $text)= @_;
	$$text=~s{<tmpl_ex(\s+[^>]+\s*)?>(.+?)</tmpl_ex[^>]*>}
	         [ &_replaceEx($1, $2, $base, $opt, $temp) ]isge;
	$$text=~m{(?:<tmpl_ex[^>]*>|</tmpl_ex[^>]*>)}
	  and croak q{At least one <TMPL_EX> not terminated at end of file!};
	$$text=~s{<tmpl_set([^>]+)>} [ &_replaceSet($1, $opt->{_ex_params}) ]isge;
}
sub _offFilter {
	my($param, $text)= @_;
	$$text=~s{<tmpl_ex\s+[^>]+\s*?>.+?</tmpl_ex[^>]*>} []isg;
	$$text=~s{(?:<tmpl_ex[^>]*>|</tmpl_ex[^>]*>)} []isg;
	$$text=~s{<tmpl_set([^>]+)>} [ &_replaceSet($1, $param) ]isge;
}
sub _replaceSet {
	my $opt  = shift || return "[ tmpl_set Error!! ]";
	my $param= shift || return "[ tmpl_set Error!! ]";
	my $name = ($opt=~/name=\s*[\'\"]?([^\s\'\"]+)/)[0]
	        || return "[ tmpl_set Error!! ]";
	my $value= ($opt=~/value=\s*[\'\"](.+?)[\'\"]/)[0]
	        || ($opt=~/value=\s*([^\s]+)/)[0]
	        || return "[ tmpl_set Error!! ('$name') ]";
	$param->{$name}= $value if $value;
	"";
}
sub _replaceEx {
	my($tag, $code, $base, $opt, $temp)= @_;
	my $escape= my $default= "";
	my($exec, %attr);
	if ($tag) {
		$attr{key_name}= lc($1)     if $tag=~/name=[\"\']?([^\s\"\']+)/;
		$attr{hidden}= 1            if $tag=~/hidden=[\"\']?([^\s\"\']+)/;
		$escape = qq{ escape="$1"}  if $tag=~/escape=[\"\']?([^\s\"\']+)/;
		$default= qq{ default="$1"} if $tag=~/default=[\"\']?([^\s\"\']+)/;
	}
	my $ident= '__$ex_'. &_get_ident_id($opt) .'$'. (++$temp->{count}). '$__';
	$code= "no strict;\n". $code if $opt->{no_strict_exec};
	eval"\$exec= sub { $code }";  ## no critic
	$attr{function}= sub { $exec->(@_) || "" };
	$opt->{_ex_orders}{$ident}= \%attr;
	$opt->{_ex_mark}{$ident}  = $ident;
	qq{<tmpl_var name="$ident"$escape$default>};
}
sub _commit_to_cache {
	my($self)= @_;
	push @{$self->{parse_stack}}, $self->{options}{_ex_orders};
	push @{$self->{parse_stack}}, $self->{options}{_ex_params};
	push @{$self->{parse_stack}}, $self->{options}{_ex_mark};
	push @{$self->{parse_stack}}, ($self->{_ex_charset} || "");
	HTML::Template::_commit_to_cache(@_);
}
sub _get_ident_id {
	$_[0]->{___ident_id} ||= substr(md5_hex(time(). {}. rand()), 0, 32);
}

package HTML::Template::Ex::DummyObject;
use strict;
sub new { bless {}, shift }

1;

__END__

=head1 NAME

HTML::Template::Ex - The Perl code is operated in the template for HTML::Template.

=head1 SYNOPSIS

  package MyProject;
  use CGI;
  use Jcode;
  use HTML::Template::Ex;
  
  my $cgi = CGI->new;
  my $self= bless { cgi=> cgi }, __PACKAGE__;
  
  my $template= <<END_OF_TEMPLATE;
  <html>
  <head><title><tmpl_var name="title"></title></head>
  <body>
  <tmpl_set name="title" value="HTML::Template::Ex">
  
  <h1><tmpl_var name="page_title"></h1>
  <h2><tmpl_var name="title"></h2>
  
  <div style="margin:10; background:#DDD;">
  <tmpl_ex>
    my($self, $param)= @_;
    $param->{page_title}= 'My Page Title';
    return $self->{cgi}->param('name') || 'It doesn't receive it.';
  </tmpl_ex>
  </div>
  
  <div style="margin:10; background:#DDD;">
  <tmpl_loop name="users">
   <div>
   <tmpl_var name="u_name" escape="html">
   : <tmpl_var name="email" escape="html">
    </div>
  </tmpl_loop>
  </div>
  
  <tmpl_ex name="users">
    return [
     { u_name=> 'foo', email=> 'foo@mydomain'    },
     { u_name=> 'boo', email=> 'boo@localdomain' },
     ];
  </tmpl_ex>
  
  <tmpl_var name="env_remote_addr">
  
  <body></html>
  END_OF_TEMPLATE
  
  my $tmpl= HTML::Template::Ex->new($self, {
    setup_env=> 1,
    scalarref=> \$template,
    encoder  => sub { Jcode->new($_[0])->euc },
    # ... other 'HTML::Template' options.
    });
  
  print STDOUT $cgi->header, $tmpl->output;

=head1 DESCRIPTION

This module offers the function to evaluate the Perl code to the template that
HTML::Template uses.

The character string enclosed with '<TMPL_EX> ... </TMPL_EX>' is evaluated as
Perl code.

  <tmpl_ex>
    my($self, $param)= @_;
    my $hoge= $self->to_method;
    $param->{hoge}= $hoge;
    return "";
  </tmpl_ex>

The object passed to the constructor is passed to the first argument to the
tmpl_ex tag.

The second argument is HASH reference for the parameter that HTML::Template uses.

When the ending value of each tmpl_ex block is returned, the value is buried
under the position.
Therefore, it is necessary to return the dead blank character to bury anything.

When HTML::Template::Ex evaluates a little code, there is a little habit when a 
complex thing is done though it is convenient.

One is the priority level when two or more tmpl_ex blocks are described.

There is no problem for one template.
It is sequentially evaluated on.
And, please pay attention to the point done earlier than HTML::Template evaluates
tag about this evaluation.

The problem reads other templates from the template.
It is time when it exists also in the read template the tmpl_ex block.
HTML::Template::Ex is not intervened for the include of the template.
Therefore, after HTML::Template evaluates tag, the tmpl_ex block include ahead
will be evaluated. This sometimes causes confusion.

  <tmpl_include name="first.tmpl">  ... [3]
  <tmpl_ex>
   .... [ 1 ]
  </tmpl_ex>
  <tmpl_include name="middle.tmpl"> ... [4]
  <tmpl_ex>
   .... [ 2 ]
  </tmpl_ex>
  <tmpl_include name="end.tmpl">    ... [5]

And, the error message is a very difficult secondarily thing.

As for each tmpl_ex block, the code is individually evaluated with eval.
Therefore, it is not a translation processed while looking about the entire 
template. Therefore, only the error where eval originates is obtained.
As for this, specific in the error generation part becomes very difficult.

The thing of this problem solved only by HTML::Template::Ex is difficult.
Therefore, the improvement is not scheduled in the future.

Using HTML::Template::Ex on the assumption of the thing to write a complex
code doesn't come recommended though it is regrettable.
Intuition is compelled every time the error occurs and debugging is compelled 
to reliance. Perhaps, this will be annoyed by the stress.

Still, I think that it can do the template that works more high-speed than
HTML::Mason and Template ToolKit if the code can be completed.
Please try and look at interesting one. 

Therefore, HTML::Template::Ex selects it by the rental server etc. of the
template driver. We will recommend the usage not to expect many of the template
driver in the situation in which the leg is limited.


And, '<TMPL_IF >' doesn't have the meaning because it is evaluated earlier than
HTML::Template though it is a thing misunderstood easily at the end.

  <tmpl_if name="hoge">
    <tmpl_ex>
      ....
    </tmpl_ex>
  <tmpl_exse>
    <tmpl_ex>
      ....
    </tmpl_ex>
  </tmpl_if>

This is smoothly evaluated to diverge to both of the tmpl_ex block.
Please solve 'IF' related to tmpl_ex block in tmpl_ex block.

  <tmpl_ex>
    my($self, $param)= @_;
    if ($apram->{hoge}) {
      ...
    } else {
      ...
    }
    "";
  </tmpl_ex>


=head1 TAGS

It is enhancing tag that introduces here added by using HTML::Template::Ex. 
Please see the document of HTML::Template about standard tag of L<HTML::Template>.

=head2 <TMPL_EX ... > [PERL_CODE] </TMPL_EX>

The Perl code is evaluated in the template.

Please see DESCRIPTION for details concerning basic operation. 

When the NAME attribute is given, the value that the tmpl_ex block returned to
the parameter of the name is substituted.
Please give the HIDDEN attribute at the same time to prevent the value being
buried under the position of the tmpl_ex block. 

  <h1><tmpl_var name="hoge"></h1>
  
  <tmpl_ex name="hoge" hidden="1">
    my($self)= @_;
    ..... ban, bo, bon.
    return $self->request->param('Fooo');
  </tmpl_ex>

The ESCAPE attribute can be used. 

  <tmpl_ex escape="html">
    ..... ban, bo, bon.
    return "<font>Zooooo</font>";
  </tmpl_ex>

* After it is escaped of html, this is buried.

=head2 <TMPL_SET NAME='...' VALUE='...'>

The value is set in the parameter in the template.

=head2 <tmpl_var name='env_*[ Environment variable name. ]'>

When the setup_env option is given to the constructor, the environment variable
is set up. 
It can be referred to by '<TMPL_VAR NAME=...>'. Please specify the environment
variable name putting up 'env_' to the head of the name.

  <tmpl_var name="env_remote_addr">
  <tmpl_var name="env_path_info">

=head1 METHODS

=head2 new ([OBJECT], [OPTION])

It is a constructor.

An arbitrary object is given to the first argument.
The given object is passed as the first argument of each tmpl_ex block.

[OPTION] is an option to pass to HTML::Template. 
Please include the option of HTML::Template::Ex here.

Options.

=over 4

=item * setup_env

It is set up to refer to the environment variable.

=item * exec_off

All the tag that HTML::Template::Ex evaluates is invalidated and it puts it out.

=item * no_strict_exec

This is turned off though the code of the tmpl_ex block is evaluated under the strict 
environment usually. For person who is not accustomed to making strict code.

=item * encoder

The CODE reference to process the character-code can be defined.

=item * auto_encoder

When charset was able to be acquired from the template, encoder is processed.

=back

=head2 charset

When charset was able to be acquired from the template, the value is returned.

=head2 other

Please refer to the document of L<HTML::Template> for other methods.

=head2 NOTES

There is causing the defective operation according to the kind of the cash used
by HTML::Template option.

* If it is 'cache' option to specify at mod_perl, it operates normally usually.

=head1 SEE ALSO

L<Egg::Release>,

=head1 AUTHOR

Masatoshi Mizuno E<lt>mizunoE<64>bomcity.comE<gt>

=head1 COPYRIGHT

Copyright (C) 2007 by Bee Flag, Corp. E<lt>L<http://egg.bomcity.com/>E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut
